### ISSUE TITLE
(JSB) CocosBuilder outlets from nested CCBs not accessible

### ISSUE THREAD

**Issue by jdmunro** [cocos2d/cocos2d-x#2856]
If you create a CCB file with another CCB nested inside, the outlets of the nested CCB are not accessible at runtime through the JavaScript bindings when the file is loaded using `cc.BuilderReader.load`.

If you log the outlets as they are loaded (`jsb_cocosbuilder.js`), e.g:

```
for (var i = 0; i < ownerOutletNames.length; i++)
{
    var outletName = ownerOutletNames[i];
    var outletNode = ownerOutletNodes[i];
    cc.log("Outlet name: " + outletName);

    owner[outletName] = outletNode;
}
```

The outlets from the nested CCB are never set on the owner variable by the loading routine, and are therefore not accessible by the JSB runtime.

EDIT: just to clarify, you can access the outlet of the nested CCB itself, but if that CCB contains any outlets, they won't be available by the runtime.

**Comment by dumganhar** [comment]
Do you have any ideas to fix it?

**Comment by jdmunro** [comment]
I'm not sure just yet - I'll see what else I can discover.

**Comment by StoneJT** [comment]
I encountered this at work about a month ago, when trying to implement automatic setting of owner node if it's not provided. 

Part of the problem is that mOwnerOutletNames of CCBReader is an instance of std::vector, and is therefore tied to a given CCBReader instance. On the other hand, mOwnerOutletNodes is a pointer to a CCArray, which in the case of sub-CCBs has been passed in the from the parent loader. Sub-CCBs will therefore actually pass outlet nodes up through the builder hierarchy, but not the names. We first noticed this when objects that were expected to be CCLabelBMFonts were actually CCSprites.

I'm not sure of the best way to go about fixing it. The simplest way would probably be to use a CCArray containing CCStrings for mOwnerOutletNames. The initial loader should then contain the names and nodes that were marked as owner variables in any sub-CCBs, and then assign them to the owner associated with the loader. 

The downsides are that all owner variables in the hierarchy are assigned to a single object (which may not be easily accessible at a later time), and that it lacks granularity. We've added an "owner" property to "owned" objects (set when outlets are set against the owner in jsb_cocosbuilder.js) to help with the first problem, and the second may actually be the intended vision of the "owner" code connection.
